### ISSUE TITLE
Using a `ParamSpec` type variable as a type parameter to both a `Generic` type variable and a `Callable` is incorrectly identified as an

### ISSUE THREAD

**Issue by favll** [microsoft/pyright#5614]
**Describe the bug**

Using a `ParamSpec` type variable as a type parameter to both a `Generic` type variable and a `Callable` is incorrectly identified as an error by pyright. Could be related to the discussion in https://github.com/microsoft/pyright/issues/5488

**Code or Screenshots**
I can reproduce the error with the following example code:

```python
from typing import Callable, Generic, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")


class MyClass(Generic[P, T]):
    func: Callable[P, T]

    def __init__(self, func: Callable[P, T]) -> None:
        self.func = func


def factory_func(func: Callable[P, T]) -> MyClass[P, T]:
    return MyClass(func=func)


def my_func(a: int, b: str) -> float:
    return a + float(b)


# This works
print(MyClass(func=my_func))

# This breaks
print(factory_func(func=my_func))
# error: Argument of type "(a: int, b: str) -> float" cannot be assigned to parameter "func" of type "(**P@factory_func) -> T@factory_func" in function "factory_func"
#    Type "(a: int, b: str) -> float" cannot be assigned to type "(**P@factory_func) -> T@factory_func" (reportGeneralTypeIssues)
# 1 error, 0 warnings, 0 informations

```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
I am running pyright as a command-line tool. Version `1.1.309` identifies 0 errors. Versions `1.1.310` to `1.1.318` (most current version) all identify an error.

**Comment by erictraut** [comment]
Thanks for the bug report. This was already fixed in last week's 1.1.319 release. It looks like the [community-maintained python wrapper](https://pypi.org/project/pyright/) is one version out of date on pypi.

**Comment by favll** [comment]
Sorry, I missed the related issue https://github.com/microsoft/pyright/issues/5495 and the new version. Just built the CLI with version 1.1.319 locally and this fixed the issue. Thank you 👍

**Comment by cainmagi** [comment]
Hello, @erictraut!

> Sorry for causing the inconvenience. I incorrectly posted the following comments to a different issue before. So I moved the comments here.

I am not sure whether the following usages are legal or not. It seems that the high-order issue of `ParamSpec` still exists in the following usages:

```python
from typing import Generic, TypeVar, Callable, Type
from typing_extensions import ParamSpec, Protocol, reveal_type


P = ParamSpec("P")
T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)


class TestProtocol(Protocol[P, T_co]):
    def __init__(self, *args: P.args, **kwargs: P.kwargs) -> None:
        ...

    def __new__(cls) -> T_co:
        ...


class Test(Generic[T]):
    def __init__(self, val1: T, val2: int) -> None:
        ...


def test2(val1: T, val2: int) -> Test[T]:
    ...


def func(obj: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


def func2(obj: Type[TestProtocol[P, T]], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


val_test = Test(val1=1, val2=2)


reveal_type(Test(val1=1, val2=2))  # Test[int], as expected.
reveal_type(test2(val1=1, val2=2))  # Test[int], as expected.
reveal_type(func(Test, val1=1, val2=2))  # Test[Unknown], however, expecting Test[int].
reveal_type(func2(Test, val1=1, val2=2))  # Test[Unknown], however, expecting Test[int].
```

This issue prevents me from defining a function accepting a "generic" instance constructor as its argument. 

By the way, I think the definition of `func2` may be problematic because it assumes the returned value of `TestProtocol.__new__` to be the returned value of the constructor. However, the type checker does not raise warnings here.

Pylance: 2023.8.10
Pyright: 932e542d

Thank you!
